scripts: edtlib: Automatically include base.yaml in all bindings

ulfalizer opened the pull request:
Main commits:

```
scripts: edtlib: Automatically include base.yaml in all bindings

This removes some boilerplate from bindings and makes things more
consistent. One downside is that all properties declared in base.yaml
can now appear on all nodes, including on nodes where they don't make
sense, but it's probably worth it.

base.yaml is moved to scripts/dts/base.yaml, and is considered part of
edtlib internals.

The automatic base.yaml inclusion is documented in
dts/binding-template.yaml and in the legacy syntax section of
doc/guides/dts/index.rst.

'include: base.yaml' is still accepted, but is ignored, and makes edtlib
print a deprecation warning. The warning mentions the new location of
base.yaml.

Remove a 'clocks' declaration from dts/bindings/clock/fixed-clock.yaml,
which declares it as 'type: array'. It clashes with the declaration for
'clocks' in base.yaml, which has it as 'type: phandle-array'. No 'clocks
= <(number) (number) ...>' assignments exist anywhere, which is the form
of assignment expected for 'type: array'.

Piggyback some binding consistency nits and a small bugfix that allows
empty files to be included (the crux is that PyYAML returns None instead
of an empty dict for them).

The empty file fix is needed because of dts/bindings/mmc/mmc.yaml, which
turns empty besides some comments if the 'include: base.yaml' is
removed. Maybe it should be removed instead.
```

```
dts: bindings: Remove base.yaml includes

base.yaml is now included automatically, so it doesn't need to be
included explicitly.
```

Piggybacked nit:

```
scripts: edtlib: Fix some *-map-mask/*-map-pass-thru typos in comments

The "map" part got lost.
```

erwango commented:
@ulfalizer, I understand the interest but I'm afraid this could confuse people. One advantage of current inclusion of base.yaml is that it makes script logic vs bindings quite transparent to users.
If we're doing this change, would be nice to keep a clear enought comment in all bindings to make base.yaml inclusion less opaque.

ulfalizer commented:
> @ulfalizer, I understand the interest but I'm afraid this could confuse people. One advantage of current inclusion of base.yaml is that it makes script logic vs bindings quite transparent to users.

Yeah, it needs to be well-documented and discoverable at least.

I've now documented it in `dts/binding-template.yaml` and in the legacy syntax section of `doc/guides/dts/index.rst`. I also added a warning to `edtlib.py` that's printed if `base.yaml` is included:

```python
for fname in fnames:
    if fname == "base.yaml":
        self._warn(
            "'include: base.yaml' in {} is deprecated and redundant. "
            "base.yaml is now automatically included in all bindings, "
            "and can be found at scripts/dts/base.yaml."
            .format(binding_path))
        continue
```

> If we're doing this change, would be nice to keep a clear enought comment in all bindings to make base.yaml inclusion less opaque.

I wonder if the added documentation and warning might make it easy enough to figure out. Feels a bit spammy to add 2-3 lines to every binding, especially once you already know it.

pabigot commented on dts/bindings/clock/fixed-clock.yaml:
```diff
@@ -21,10 +21,5 @@ properties:
       description: output clock frequency (Hz)
       required: true
 
-    clocks:
```
Is this supposed to be removed?

ulfalizer replied:
Yup, forgot to explain it. See the updated commit message for the commit that also has the code changes (it needs to go in with that one).

Guessing it was supposed to be `type: phandle-array` (currently you'd have to assign it like `clocks = <1 2 3>`, whatever that means), and it's already in `base.yaml`. Probably close enough in meaning too.

pabigot reviewed (changes requested):
Please split the commit so the mass changes that are supposed to have no effect are in a single commit separate from anything that has intent behind it.  The piggy-back changes should be grouped by intent as well.

Verify that the `clocks` property was intentionally removed; if so, please make that its own commit with explanation, because it looks wrong.

ulfalizer committed:
dts: bindings: Remove base.yaml includes

base.yaml is now included automatically, so it doesn't need to be
included explicitly.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
scripts: edtlib: Fix some *-map-mask/*-map-pass-thru typos in comments

The "map" part got lost.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
scripts: edtlib: Automatically include base.yaml in all bindings

This removes some boilerplate from bindings and makes things more
consistent. One downside is that all properties declared in base.yaml
can now appear on all nodes, including on nodes where they don't make
sense, but it's probably worth it.

base.yaml is moved to scripts/dts/base.yaml, and is considered part of
edtlib internals.

The automatic base.yaml inclusion is documented in
dts/binding-template.yaml and in the legacy syntax section of
doc/guides/dts/index.rst.

'include: base.yaml' is still accepted, but is ignored, and makes edtlib
print a deprecation warning. The warning mentions the new location of
base.yaml.

Remove a 'clocks' declaration from dts/bindings/clock/fixed-clock.yaml,
which declares it as 'type: array'. It clashes with the declaration for
'clocks' in base.yaml, which has it as 'type: phandle-array'. No 'clocks
= <(number) (number) ...>' assignments exist anywhere, which is the form
of assignment expected for 'type: array'.

Piggyback some binding consistency nits and a small bugfix that allows
empty files to be included (the crux is that PyYAML returns None instead
of an empty dict for them).

The empty file fix is needed because of dts/bindings/mmc/mmc.yaml, which
turns empty besides some comments if the 'include: base.yaml' is
removed. Maybe it should be removed instead.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
> Please split the commit so the mass changes that are supposed to have no effect are in a single commit separate from anything that has intent behind it. The piggy-back changes should be grouped by intent as well.

Split up with (most) of the binding changes separately now.

dbkinder commented on dts/binding-template.yaml:
```diff
@@ -27,18 +27,38 @@ compatible: "manufacturer,device"
-# 'properties' description below).
+# The file scripts/dts/base.yaml is automatically included in all bindings. It
+# defines generic properties that appear on many devicetree nodes. It is a good
+# to idea to look at it to get an idea of what definitions are already
```
``It is a good idea ...`` (delete ``to``)

dbkinder commented on dts/binding-template.yaml:
```diff
@@ -27,18 +27,38 @@ compatible: "manufacturer,device"
+# available.
+#
+# Included files are merged into bindings with a simple recursive dictionary
+# merge. It is up the binding author to make sure that the resulting binding is
```
``It is up to the ...``  (add ``to``)

ulfalizer replied:
Fixed

dbkinder reviewed (changes requested):
fixing typos in the binding-template.yaml file

erwango commented:
@ulfalizer I guess there are a couple of on-going PR which will have a silent conflict with this change. What is the output if base.yaml is actually included in a binding?

ulfalizer commented:
> @ulfalizer I guess there are a couple of on-going PR which will have a silent conflict with this change. What is the output if base.yaml is actually included in a binding?

Get this warning:

```
warning: 'include: base.yaml' in some/binding.yaml is deprecated and redundant. base.yaml is
now automatically included in all bindings, and can be found at scripts/dts/base.yaml.
```

Other than that it's a no-op (the code specifically ignores `base.yaml`), and `scripts/dts/base.yaml` still gets included implicitly, so it won't affect the output.

galak reviewed (changes requested):
I'm not sure I think this makes sense.  I get the value in reducing the need to include base.yaml everywhere, but I think its better to be explicit vs having an implied "inclusion".

ulfalizer commented:
> I'm not sure I think this makes sense. I get the value in reducing the need to include base.yaml everywhere, but I think its better to be explicit vs having an implied "inclusion".

One nice thing about it besides shortening bindings is that it forces consistency. If someone tries to add a property to a binding that clashes with something in `base.yaml`, it will be an error. That revealed an issue in `fixed-clock.yaml` in this PR, for example.

Some other systems (e.g. `bitbake` and `dt-schema`) also have base configuration files that are included implicitly.

It makes things slightly less transparent, but I think people will pick up on it pretty quickly.

ulfalizer commented:
Should also speed up binding parsing a bit btw, because `base.yaml` is only parsed once now.

dbkinder reviewed (approved):
LGTM, thanks.

erwango commented:
@galak, should we keep this one ?
